File: test/unit/widgets/gamification/resolve_nudge_label_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:snabbit_runner/models/gamification/nudge_label.dart';
import 'package:snabbit_runner/providers/language_provider.dart';
import 'package:snabbit_runner/widgets/gamification/resolve_nudge_label.dart';

void main() {
  group('resolveNudgeLabel', () {
    final lang = LanguageProvider();

    test('default_text substitutes params e.g. {{time}}', () {
      expect(
        resolveNudgeLabel(
          NudgeLabel(
            key: 'nudge_early_login_avoid_no_show',
            params: const {'time': '08:00 am', 'red_cards': 50},
            defaultText: 'Login by {{time}} to avoid No Show',
          ),
          lang,
        ),
        'Login by 08:00 am to avoid No Show',
      );
    });

    test('sheet warning keys use English defaults when not in bundle', () {
      expect(
        resolveNudgeLabel(
          const NudgeLabel(key: 'nudge_mark_present_earn', params: {}),
          lang,
        ),
        '**Mark present** to earn for the jobs you complete',
      );
      expect(
        resolveNudgeLabel(
          const NudgeLabel(key: 'nudge_provisional_absent', params: {}),
          lang,
        ),
        '**Mark absent** — confirm to continue',
      );
    });

    test('nudge_false_attendance_penalty substitutes redCards when present', () {
      expect(
        resolveNudgeLabel(
          const NudgeLabel(
            key: 'nudge_false_attendance_penalty',
            params: {'redCards': 2},
          ),
          lang,
        ),
        '**False attendance** — **2** red cards penalty',
      );
    });

    test('nudge_false_attendance_penalty omits placeholder when redCards absent', () {
      expect(
        resolveNudgeLabel(
          const NudgeLabel(key: 'nudge_false_attendance_penalty', params: {}),
          lang,
        ),
        '**False attendance** — red card penalty may apply',
      );
    });
  });
}

File: shared/src/commonTest/kotlin/com/snabbit/runner/shared/features/gamification/presentation/ResolveNudgeLabelTest.kt
package com.snabbit.runner.shared.features.gamification.presentation

import com.snabbit.runner.shared.features.gamification.domain.model.NudgeLabel
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ResolveNudgeLabelTest {

    @Test
    fun legacyLiteral_returnsText() {
        val label = NudgeLabel.literal("Just do it")
        assertEquals("Just do it", ResolveNudgeLabel.resolve(label))
    }

    @Test
    fun defaultText_substitutesParams() {
        val label = NudgeLabel(
            key = "nudge_early_login_avoid_no_show",
            params = mapOf("time" to "04:00 pm"),
            defaultText = "Login by {{time}} to avoid No Show",
        )
        assertEquals("Login by 04:00 pm to avoid No Show", ResolveNudgeLabel.resolve(label))
    }

    @Test
    fun noDefaultText_usesEnglishFallback_keepsBoldMarkers() {
        val label = NudgeLabel(key = "nudge_long_distance_bonus")
        assertEquals("**Accept long distance job** to earn", ResolveNudgeLabel.resolve(label))
    }

    @Test
    fun falseAttendance_withRedCards_substitutes() {
        val label = NudgeLabel(
            key = "nudge_false_attendance_penalty",
            params = mapOf("redCards" to "3"),
        )
        assertEquals("**False attendance** — **3** red cards penalty", ResolveNudgeLabel.resolve(label))
    }

    @Test
    fun falseAttendance_missingRedCards_usesSafeFallback() {
        val label = NudgeLabel(key = "nudge_false_attendance_penalty")
        assertEquals(
            "**False attendance** — red card penalty may apply",
            ResolveNudgeLabel.resolve(label),
        )
    }

    @Test
    fun unknownKey_withoutDefault_isEmpty() {
        assertTrue(ResolveNudgeLabel.resolve(NudgeLabel(key = "nudge_brand_new")).isEmpty())
    }
}
